snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java

```diff
@@ -186,8 +189,9 @@ public boolean schemaExists(SnowflakeIdentifier schema) {
                   queryHarness.query(
                       conn, finalQuery, TABLE_RESULT_SET_HANDLER, schema.toIdentifierString()));
     } catch (SQLException e) {
-      if (e.getErrorCode() == 2003 && e.getMessage().contains("does not exist")) {
-        return false;
+      // 2003, 2043, 2001 are possible Snowflake error codes for Object does not exist
+      if (e.getErrorCode() == 2003 || e.getErrorCode() == 2043 || e.getErrorCode() == 2001) {
+        throw new NoSuchNamespaceException("%s", e.getMessage());
```

Comment:
`schemaExists` should continue to return false instead of throwing. Throwing is just for the other methods that actually try to access the namespace or table.